x/ecocredit/server/orderfill/manager.go

```diff
@@ -0,0 +1,196 @@
+type TransferManager interface {
+	// SendCoins sends the specified amount of denom.
+	SendCoins(denom string, amount sdk.Int, from, to sdk.AccAddress) error
+
+	// SendCredits sends the specified amount of credits. If retire is true, the
+	// credits will be retired with the provided retirement location, otherwise
+	// they will be transferred and remain tradable.
+	SendCredits(batchId uint64, amount math.Dec, from, to sdk.AccAddress, retire bool, retirementLocation string) error
+}
+
+func (t manager) Fill(
+	ctx context.Context,
+	market *marketplacev1beta1.Market,
+	buyOrder *marketplacev1beta1.BuyOrder,
+	sellOrder *marketplacev1beta1.SellOrder,
+) (Status, error) {
+	retire := true
+	if buyOrder.DisableAutoRetire {
+		if !sellOrder.DisableAutoRetire {
+			return NotFilled, ecocredit.ErrInvalidBuyOrder.Wrapf("unexpected: auto-retire setting doesn't match, these orders should have never been matched")
+		}
+		retire = false
+	} else {
+		if buyOrder.RetirementLocation == "" {
+			return NotFilled, ecocredit.ErrInvalidBuyOrder.Wrapf("unexpected: retirement location is empty, this order never should have never been created.")
+		}
+	}
+
+	buyQuant, err := math.NewPositiveDecFromString(buyOrder.Quantity)
+	if err != nil {
+		return NotFilled, err
+	}
+
+	sellQuant, err := math.NewPositiveDecFromString(sellOrder.Quantity)
+	if err != nil {
+		return NotFilled, err
+	}
+
+	settlementPrice, err := math.NewPositiveDecFromString(buyOrder.BidPrice)
+	if err != nil {
+		return NotFilled, err
+	}
+
+	cmp := buyQuant.Cmp(sellQuant)
+
+	var actualQuant math.Dec
+	var status Status
+	if cmp < 0 {
+		actualQuant = buyQuant
+		status = BuyFilled
+
+		newSellQuant, err := sellQuant.Sub(buyQuant)
+		if err != nil {
+			return NotFilled, err
+		}
+		sellOrder.Quantity = newSellQuant.String()
+		err = t.marketplaceStore.SellOrderStore().Update(ctx, sellOrder)
+		if err != nil {
+			return NotFilled, err
+		}
+
+		err = t.marketplaceStore.BuyOrderStore().Delete(ctx, buyOrder)
+		if err != nil {
+			return NotFilled, err
+		}
+	} else if cmp == 0 {
+		actualQuant = buyQuant
+		status = BothFilled
+
+		err = t.marketplaceStore.SellOrderStore().Delete(ctx, sellOrder)
+		if err != nil {
+			return NotFilled, err
+		}
+
+		err = t.marketplaceStore.BuyOrderStore().Delete(ctx, buyOrder)
+		if err != nil {
+			return NotFilled, err
+		}
+	} else {
+		actualQuant = sellQuant
+		status = SellFilled
+
+		err = t.marketplaceStore.SellOrderStore().Delete(ctx, sellOrder)
+		if err != nil {
+			return NotFilled, err
+		}
+
+		newBuyQuant, err := buyQuant.Sub(sellQuant)
+		if err != nil {
+			return NotFilled, err
+		}
+		buyOrder.Quantity = newBuyQuant.String()
+		err = t.marketplaceStore.BuyOrderStore().Update(ctx, buyOrder)
+		if err != nil {
+			return NotFilled, err
+		}
+	}
+
+	err = t.transferManager.SendCredits(sellOrder.BatchId, actualQuant, sellOrder.Seller, buyOrder.Buyer, retire, buyOrder.RetirementLocation)
+
+	// TODO correct decimal precision
```

Comment:
is this being added in a later PR? tracking issue?

Reply:
no I should do it in this PR... I actually started working on this, but then didn't finish. sorry...